posts/2021-07-28-GATs-stabilization-push.md

```diff
@@ -0,0 +1,132 @@
+
+
+Okay. Phew. That's the main point of this post and the most exciting news. But as I said before, I think it's also reasonable for me to explain *what* this feature is, what you can do with it, and some of the background and how we got here.
+
+## So what are GATs?
+
+**Note: this will only be a brief overview. The [RFC] contains many more details.**
+
+GATs (generic associated types) were originally proposed in [RFC 1598][RFC]. Put simply, they allow you to define type, lifetime, or const generics on associated types. If you're familiar with languages that have "higher-kinded types", then you could calls GATs *type constructors on traits*.
+
+To jump into an example of GATs, here is a popular use case: a `LendingIterator` (formerly known as a `StreamingIterator`):
+
+```rust
+trait LendingIterator {
+    type Item<'a>;
+    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
+}
+```
+
+With this, you could create a line iterator over a `String` without cloning each line...
+```rust
+struct LineIterator { ... }
+impl LendingIterator for LineIterator {
+    type Item<'a> = &'a str;
+    ...
+}
+fn lines(text: String) -> LineIterator { ... }
+// later...
+let all_lines = lines(String::from("A\nmulti\nline\nString!\n"));
+let first_line: &str = all_lines.next();
+assert_eq!(first_line, "A");
+```
+
+Or you could create an iterator over overlapping subslices of an array...
+```rust
+struct SubSliceIterator<T> { ... }
+impl<T> LendingIterator for SubSliceIterator<T> {
+    type Item<'a> = &'a [T];
+}
+```
+
+As another example, imagine you wanted a struct to be generic over a pointer to a specific type. You might write the following code:
+
+```rust
+trait PointerFamily {
+    type Pointer<T>: Deref<Target = T>;
+    fn new<T>(value: T) -> Self::Pointer<T>;
+}
+
+struct ArcFamily;
+struct RcFamily;
+
+impl PointerFamily for ArcFamily {
+    type Pointer<T> = Arc<T>;
+    ...
+}
+impl PointerFamily for RcFamily {
+    type Pointer<T> = Rc<T>;
+    ...
+}
+
+struct MyStruct<P: PointerFamily> {
+    pointer: P::Pointer<String>,
+}
+```
+
+We won't go in-depth on the detail here, but this example is nice in that it not only highlights the use of types in GATs, but also showcases the use of the trait bounds (as you can already do on associated types).
+
+These two examples only scratch the surface of the patterns that GATs support. If you find any that seem particularly interesting or clever, we would love to hear about them over on [Zulip]!
+
+## Why has it taken so long to implement this?
+
+So why has this *really* taken nearly four years to get to the point that we are now? Well, it's hard to put into words how much the existing trait solver has had to change and adapt; but, consider this: for a while, it was thought that to support GATs, we would have transition rustc to use Chalk (which is experimental and uses logical predicates to solve trait goals) as its trait solver.
```

Comment:
Link to [chalk](https://github.com/rust-lang/chalk)? I would also describe Chalk as "a potential future trait solver" for rustc in addition to it being experimental and such.

Reply:
Oops. I made the link at the bottom but didn't actually link it. And changed the wording a bit.